### PR TITLE
Improve App Dev Guide

### PR DESCRIPTION
**Pull request by ojalatodd**
The current app dev guide flow is confusing.
This commit changes the structure and separate the docker
and ubuntu install and 'getting started' sections.

Signed-off-by: Todd Ojala <todd@bitwise.io>

### CODE REVIEW

File: docs/source/app_developers_guide/docker.rst [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+The Docker Compose file (*sawtooth-default.yaml*) starts the following
+transaction processors automatically:
+
+* tp_config
+* tp_intkey_python
+* tp_xo_python
+
+These transaction processors handle transactions for the config, intkey, and XO transaction families, and are written in Python. 
+
+Sawtooth includes additional transaction processors.
```
Comment (vaporos) [inline docs/source/app_developers_guide/docker.rst]: Let's not maintain this list here.  There are not docker containers published for all of these yet.

File: docs/source/app_developers_guide/docker.rst [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+  [19:31:06 INFO    handler] processing: Verb=set Name=HOUUQS Value=10140 address=1cf126380fa9e716a05ac815741fd1960d5952e60f8747e13334f79504c57d0287b77cf9b78284d0e1544f6f0366d66c6e6eb99dc5c154b84175b2d20008d721c7b623
+  [19:31:06 INFO    handler] processing: Verb=set Name=lrnuDC Value=92318 address=1cf12617c797cf8c27254bbdb5c9bda09f9405b9494ae32b79b9b6d30881ca8552d5932a68f703d1b6754b9feb2edafa76a797fc0826110381b0f8614f2c6853316b47
+  [19:31:06 INFO    handler] processing: Verb=set Name=BKaiql Value=94175 address=1cf12669cbc17d076a1accb4b0bb61f40ed4f999173b90e3ca2591875a55fee2947661e60fa1c57b41ef0f2660176b945a01c85ff645543297068a3fb1306324a19612
+  [19:31:06 INFO    handler] processing: Verb=set Name=wpMQmE Value=47316 address=1cf1260f6bdf66b65ff7c00ec58c4deccffd167bfee7a85698880dfa485df3de1ec18a5b2d1dc12849743d1c74320108360a2d40d223b35fbc1c4ea03bbd8306480c62
+  [19:31:06 INFO    handler] processing: Verb=set Name=GTgrvP Value=31921 address=1cf12606ac7db03c756133c07d7d02b59f3ef9eae6774fe59c75c88ab66a9fabbbaef9975dbf9aa197d1090ed126d7b18e2
+
+
+Config Transaction Family Usage
```
Comment (vaporos) [inline docs/source/app_developers_guide/docker.rst]: It should be possible to submit settings from the client container.  Since the user has already been show the client container and sawtooth command, does this need to be docker specific?

File: docs/source/app_developers_guide/docker.rst [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+
+The output of the command will be similar to this:
+
+.. code-block:: console
+
+  DATA: "b'\xa1fcCTdcH\x192B'"
+  HEAD: "0c4364c6d5181282a1c7653038ec9515cb0530c6bfcb46f16e79b77cb524491676638339e8ff8e3cc57155c6d920e6a4d1f53947a31dc02908bcf68a91315ad5"
+
+
+Next Steps
```
Comment (vaporos) [inline docs/source/app_developers_guide/docker.rst]: Remove this section. The CLI reference is to the left in the HTML output or a completely separate document in the PDF version.

File: docs/source/app_developers_guide/docker.rst [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+  1cf1260cd1c2492b6e700d5ef65f136051251502e5d4579827dc303f7ed76ddb7185a19be0c6443503594c3734141d2bdcf5748a2d8c75541a8e568bae063983ea27b9 11   b'\xa1frdLONu\x...
+  1cf126ed7d0ac4f755be5dd040e2dfcd71c616e697943f542682a2feb14d5f146538c643b19bcfc8c4554c9012e56209f94efe580b6a94fb326be9bf5bc9e177d6af52 11   b'\xa1fAUZZqk\x...
+  1cf126c46ff13fcd55713bcfcf7b66eba515a51965e9afa8b4ff3743dc6713f4c40b4254df1a2265d64d58afa14a0051d3e38999704f6e25c80bed29ef9b80aee15c65 11   b'\xa1fLvUYLk\x...
+  1cf126c4b1b09ebf28775b4923e5273c4c01ba89b961e6a9984632612ec9b5af82a0f7c8fc1a44b9ae33bb88f4ed39b590d4774dc43c04c9a9bd89654bbee68c8166f0 13   b'\xa1fXHonWY\x...
+  1cf126e924a506fb2c4bb8d167d20f07d653de2447df2754de9eb61826176c7896205a17e363e457c36ccd2b7c124516a9b573d9a6142f031499b18c127df47798131a 13   b'\xa1foWZXEz\x...
+  1cf126c295a476acf935cd65909ed5ead2ec0168f3ee761dc6f37ea9558fc4e32b71504bf0ad56342a6671db82cb8682d64689838731da34c157fa045c236c97f1dd80 13   b'\xa1fadKGve\x...
+
+
+
+Viewing Data In A Node
```
Comment (vaporos) [inline docs/source/app_developers_guide/docker.rst]: Instead of 'node', use 'address'.  'Viewing Data at an Address' and below 'copy the address you want to view'.

File: docs/source/app_developers_guide/ubuntu.rst [hunk 5]
```diff
@@ -1,41 +1,53 @@
-
-.. _intro_to_sawtooth:
-
-*****************************************
-Getting Started with Hyperledger Sawtooth
-*****************************************
-
+****************************
+Installation on Ubuntu 16.04
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: In the TOC this is "Using Sawtooth on Ubuntu 16.04"

File: docs/source/app_developers_guide/ubuntu.rst [hunk 6]
```diff
@@ -1,41 +1,53 @@
...
+****************************
 
 Overview
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: Remove the section, let the intro text flow from the parent section.

File: docs/source/app_developers_guide/ubuntu.rst [hunk 7]
```diff
@@ -1,41 +1,53 @@
...
 ========
 
-This tutorial shows you how to start a validator and transaction processor,
-then perform some basic operations, such as:
+This tutorial guides you through the process of setting up Hyperledger Sawtooth
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: Not necessarily a tutorial.

File: docs/source/app_developers_guide/ubuntu.rst [hunk 8]
```diff
@@ -1,41 +1,53 @@
...
+for application development on Ubuntu Linux, and then introduces basic Sawtooth
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: Should be "Ubuntu" not "Ubuntu Linux".

File: docs/source/app_developers_guide/ubuntu.rst [hunk 9]
```diff
@@ -1,41 +1,53 @@
...
+concepts necessary for application development. After this tutorial, you will be
+ready to perform app development tasks, such as implementing business logic with
+transaction families and writing clients which use Sawtooth's REST API.
 
-* Submitting transactions
-* Setting configuration settings using the config transaction family
-* Viewing blocks, transactions and state
+If you want to install Sawtooth using Docker Compose rather than Docker Compose, see :doc:`docker`.
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: Remove, same reason as comment on similar docker section.

File: docs/source/app_developers_guide/ubuntu.rst [hunk 10]
```diff
@@ -1,41 +1,53 @@
...
+If you want to install Sawtooth using Docker Compose rather than Docker Compose, see :doc:`docker`. 
 
-If you have not yet set your environment up, see :doc:`environment_setup` in
-the App Developer's Guide.
 
 Prerequisites
--------------
+=============
+
+Sawtooth requires Ubuntu version 16.04 or higher.
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: Remove this section.  The overall content is for "Using Sawtooth on Ubuntu 16.04".  We do not support any other Ubuntu version with our current packaging.  Sawtooth itself might work though, if installed via some other method, so "Sawtooth requires" is not accurate.

File: docs/source/app_developers_guide/ubuntu.rst [hunk 11]
```diff
@@ -1,41 +1,53 @@
...
+
+
+Installation
+============
+
+You can install Sawtooth directly on your Ubuntu machine, using the
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: Simplify "You can install Sawtooth on Ubuntu using the following steps:".  "Ubuntu machine" sounds like physical hardware, and this may be a virtual machine.

File: docs/source/app_developers_guide/ubuntu.rst [hunk 12]
```diff
@@ -1,41 +1,53 @@
...
+
+.. code-block:: console
+
+  $ echo "deb http://repo.sawtooth.me/ubuntu/0.8/stable xenial universe" >> /etc/apt/sources.list
+  $ apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 6B58B1AC10FB5F63
+  $ apt-get update && apt-get install -y sawtooth
+
+
+
+Running The Validator As A Service
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: Remove empty section.

File: docs/source/app_developers_guide/ubuntu.rst [hunk 13]
```diff
@@ -47,33 +59,37 @@ test networks. In this case, you need to create a genesis block when
...
+shown below using the **sawtooth config genesis** command. 
 
 To create the genesis block, log in to the development environment CLI and run
 the following command:
 
 .. code-block:: console
 
-  $ sawtooth admin genesis
-  Generating /home/ubuntu/sawtooth/data/genesis.batch
+  $ sawtooth keygen my_key
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: Omit my_key and the -k option on config genesis below.  This is kind of advanced usage, but here we are introducing the sequence, so we should simplify it.

File: docs/source/app_developers_guide/ubuntu.rst [hunk 14]
```diff
@@ -137,6 +144,18 @@ The transaction processor produces the following output:
 To stop the transaction processor, press CTRL-c.
 
 
+Starting the Rest API
```
Comment (vaporos) [inline docs/source/app_developers_guide/ubuntu.rst]: "REST API" instead of "Rest API"

File: docs/source/app_developers_guide/ubuntu.rst (hunk 5 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: Fixed.

File: docs/source/app_developers_guide/ubuntu.rst (hunk 7 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: My change: tutorial-->document

File: docs/source/app_developers_guide/ubuntu.rst (hunk 8 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: fixed

File: docs/source/app_developers_guide/ubuntu.rst (hunk 9 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: Removed

File: docs/source/app_developers_guide/ubuntu.rst (hunk 10 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: Removed

File: docs/source/app_developers_guide/ubuntu.rst (hunk 11 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: Done

File: docs/source/app_developers_guide/ubuntu.rst (hunk 12 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: Filled this in and making it useful

File: docs/source/app_developers_guide/ubuntu.rst (hunk 13 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: Simplified and tested.

File: docs/source/app_developers_guide/ubuntu.rst (hunk 14 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: Fixed

File: docs/source/app_developers_guide/docker.rst [hunk 15]
```diff
@@ -0,0 +1,459 @@
+********************
+Using Docker Compose
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: fixed

File: docs/source/app_developers_guide/docker.rst [hunk 16]
```diff
@@ -0,0 +1,459 @@
+********************
+Using Docker Compose 
+********************
+
+Overview
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Removed header. Did you mean just let the intro live under the document title (Using Sawtooth with Docker)? That is what I did, and I'll also do the same for the Ubuntu guide.

File: docs/source/app_developers_guide/docker.rst [hunk 17]
```diff
@@ -0,0 +1,459 @@
...
+========
+
+This tutorial guides you through the process of setting up Hyperledger Sawtooth
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: tutorial-->document

File: docs/source/app_developers_guide/docker.rst [hunk 18]
```diff
@@ -0,0 +1,459 @@
...
+for application development using Docker Compose, and then introduces basic
+Sawtooth concepts necessary for application development. After this tutorial,
+you will be ready to perform app development tasks, such as implementing
+business logic with transaction families and writing clients which use
+Sawtooth's REST API.
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: OK, I reworked this section, but did leave a teaser in on the more advanced stuff. I think this is important from a marketing POV.

File: docs/source/app_developers_guide/docker.rst [hunk 19]
```diff
@@ -0,0 +1,459 @@
...
+
+If you want to install Sawtooth on Ubuntu Linux rather than Docker Compose, see
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Removed

File: docs/source/app_developers_guide/docker.rst [hunk 20]
```diff
@@ -0,0 +1,459 @@
...
+:doc:`ubuntu`.
+
+
+Prerequisites
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Renamed. I think the suggestion is good.

File: docs/source/app_developers_guide/docker.rst [hunk 21]
```diff
@@ -0,0 +1,459 @@
...
+=============
+
+The following tools are required:
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Redid this section.

File: docs/source/app_developers_guide/docker.rst [hunk 22]
```diff
@@ -0,0 +1,459 @@
...
+
+* `Docker Engine <https://docs.docker.com/engine/installation/>`_ (17.03.0-ce
+  or later)
+* `Docker Compose <https://docs.docker.com/compose/install/>`_
+	
+	- On Windows and macOS, Docker Compose is installed automatically when you install Docker Engine
+
+Choosing the Right Version of Docker
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Removed

File: docs/source/app_developers_guide/docker.rst [hunk 23]
```diff
@@ -0,0 +1,459 @@
...
+------------------------------------
+
+.. todo::
+
+	Info on choosing the correction version of docker
+
+Supported Windows Versions
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Removed.

File: docs/source/app_developers_guide/docker.rst [hunk 24]
```diff
@@ -0,0 +1,459 @@
...
+
+
+Installation
+============
+
+Download the Docker Compose File
+--------------------------------
+
+A docker compose file is provided which defines the process for constructing
+a sawtooth environment, including the container images to download and the
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed

File: docs/source/app_developers_guide/docker.rst [hunk 25]
```diff
@@ -0,0 +1,459 @@
...
+First, install curl on the client container, by running the following commands 
+from the client container terminal window:
+
+.. code-block:: console
+
+	$ sudo sed -i -e 's/us.archive.ubuntu.com/archive.ubuntu.com/g' /etc/apt/sources.list
+	$ apt-get update
+	$ apt-get install curl
+
+You can now use the curl command to test the connectivity betweent the client container
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: fixed

File: docs/source/app_developers_guide/docker.rst (hunk 1 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Removed

File: docs/source/app_developers_guide/docker.rst (hunk 2 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: The commands above were submitted in the client container, according to the directions above:
"Run the following commands from the client container:"

Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: However, I did redo the section a bit, with actual output from the docker containers.

File: docs/source/app_developers_guide/docker.rst (hunk 4 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed

File: docs/source/app_developers_guide/docker.rst (hunk 3 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Also removed from Ubuntu guide.

File: docs/source/app_developers_guide/docker.rst [hunk 26]
```diff
@@ -0,0 +1,459 @@
...
+  this process, see :doc:`ubuntu`.
+
+
+Confirming Connectivity 
+=======================
+
+To confirm that a validator is up and running, and reachable from the client container, 
+you can use the curl command. 
+
+First, install curl on the client container, by running the following commands
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: I'll address this tomorrow with a bit of advice from @rbuysse

**Comment by ojalatodd** [comment]
@vaporos  I made all the changes.
Log locations are not yet described. Will add that info in the next Sprint.

File: docs/source/app_developers_guide/docker.rst [hunk 27]
```diff
@@ -0,0 +1,417 @@
...
+  validator_1         | [19:03:47.538 INFO     cli] config [path]: config_dir = "/etc/sawtooth"
+  tp_intkey_python_1  | [19:03:47 DEBUG   selector_events] Using selector: ZMQSelector
+  validator_1         | [19:03:47.538 INFO     cli] config [path]: key_dir = "/etc/sawtooth/keys"
+  tp_intkey_python_1  | [19:03:47 INFO    core] register attempt: OK
+
+Open a new terminal so we can connect to the client container:
+
+.. code-block:: console
+
+  % docker exec -it compose_client_1 bash
```
Comment (rbuysse) [inline docs/source/app_developers_guide/docker.rst]: Right now, the container names from docker-compose are based on the folder where the compose file is located and thus aren't predictable. I submitted a PR to change this behavior:

https://github.com/hyperledger/sawtooth-core/pull/572

File: docs/source/app_developers_guide/docker.rst [hunk 28]
```diff
@@ -0,0 +1,417 @@
...
+  validator_1         | [21:02:53.195 DEBUG    tp_state_handlers] GET: [('1cf126721fff0dc4ccb345fb145eb9e30cb7b046a7dd7b51bf7393998eb58d40df5f9a', b'\xa1fZeuYwh\x1a\x00\x019%')]
+  validator_1         | [21:02:53.200 DEBUG    interconnect] ServerThread sending TP_STATE_GET_RESPONSE to b'63cf2e2566714070'
+  tp_intkey_python_1  | [21:02:53.202 DEBUG    handler] Incrementing "ZeuYwh" by 1
+
+
+Config Transaction Family Usage
+===============================
+
+Sawtooth provides a :doc:`config transaction family
+<../transaction_family_specifications/config_transaction_family>` that stores on-
```
Comment (rbuysse) [inline docs/source/app_developers_guide/docker.rst]: This is rendering weird in the generated docs. Can you move the "on-" to the line below?

File: docs/source/app_developers_guide/docker.rst [hunk 29]
```diff
@@ -0,0 +1,417 @@
...
+.. note::
+
+  The config command needs to use a key generated in the validator container.
+  Thus, you must open a terminal window running in the validator container,
+  rather than the client container (for the following command only).
+  Run the following command from your host machine's CLI:
+
+.. code-block:: console
+
+  % docker exec -it compose_validator_1 bash
```
Comment (rbuysse) [inline docs/source/app_developers_guide/docker.rst]: Same as my note above, we should merge #572 and then update the container names.

File: docs/source/app_developers_guide/ubuntu.rst [hunk 30]
```diff
@@ -1,42 +1,69 @@
...
+the following components, which can then be started, stopped, and restarted
+using the *systemctl* command:
+
+* validator
+* transaction processors
+* rest_api
 
-.. caution::
+To run sawtooth components as a service, replace the manual commands in the next
+sections with a *systemctl* command.
```
Comment (rbuysse) [inline docs/source/app_developers_guide/ubuntu.rst]: If the components are started using systemctl, all of the config data and logs need to be readable and writeable by the 'sawtooth' user which is automatically created by a post-install script in the validator deb.

I'm not sure if the right pattern should be to become the sawtooth user during configuration or to execute the commands as root and then chown the files later.

File: docs/source/app_developers_guide/docker.rst (hunk 28 above)
Comment (rbuysse) [inline docs/source/app_developers_guide/docker.rst]: There's one of these in the ubuntu doc at https://github.com/hyperledger/sawtooth-core/pull/546/files#diff-5f7497f925dbae7a0d4e62bcc9df1242R284. Since you're already in the docs, can you fix that one too?

File: docs/source/app_developers_guide/docker.rst [hunk 31]
```diff
@@ -0,0 +1,417 @@
...
+
+Download the docker compose file `here <./sawtooth-default.yaml>`_.
+
+Environment Startup
+-------------------
+
+To start up the environment, perform the following tasks:
+
+1. Open a terminal window.
+2. Change your working directory to the same directory where you saved the`
```
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed.

File: docs/source/app_developers_guide/docker.rst (hunk 27 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Replacing this text block with output from PR #572

File: docs/source/app_developers_guide/docker.rst (hunk 28 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed. I'll do a test build, also.

File: docs/source/app_developers_guide/docker.rst (hunk 29 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Updating,

File: docs/source/app_developers_guide/ubuntu.rst (hunk 30 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/ubuntu.rst]: Do we need to address this for this version? What's the best idea to add something to this document for this sprint?

File: ci/sawtooth-all [hunk 32]
```diff
@@ -33,6 +33,7 @@ RUN echo "deb http://repo.sawtooth.me/ubuntu/0.8/stable xenial universe" >> /etc
  && apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 6B58B1AC10FB5F63 \
  && apt-get update \
  && apt-get install -y -q \
+    curl \
```
Comment (vaporos) [inline ci/sawtooth-all]: This should really be in a separate commit.

File: docs/source/app_developers_guide/docker.rst (hunk 27 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: confirmed I got the right name

File: docs/source/app_developers_guide/docker.rst (hunk 29 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed with the new container name

File: ci/sawtooth-all (hunk 32 above)
Comment (ojalatodd) [inline ci/sawtooth-all]: Noted. Sorry!

File: docs/source/app_developers_guide/docker.rst [hunk 33]
```diff
@@ -0,0 +1,416 @@
...
+depicted.
+
+
+Installing Docker Engine and Docker Compose
+===========================================
+
+On Windows and macOS, install the latest version of `Docker Engine <https://docs.docker.com/engine/installation/>`_.
+
+  - On Windows and macOS, Docker Compose is installed automatically when you
+  - install Docker Engine
```
Comment (rbuysse) [inline docs/source/app_developers_guide/docker.rst]: extra "-" here.

Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed

File: docs/source/_templates/sdk_TP_tutorial.rst [hunk 34]
```diff
@@ -41,12 +41,13 @@ A full implementation of the tic-tac-toe transaction family can be found in
...
 
-This tutorial assumes that you have gone through :doc:`/app_developers_guide/intro_to_sawtooth` and are
+This tutorial assumes that you have gone through :doc:`/app_developers_guide/getting_started` and are
 familiar with the concepts introduced there.
 
 Prior to going through this tutorial, you should have a working Sawtooth
 environment.  Specific setup instructions are available
-in :doc:`/app_developers_guide/environment_setup`.
+in :doc:`/app_developers_guide/getting_started`, with separate guides
+for Docker Compose and Ubuntu Linux.
```
Comment (aludvik) [inline docs/source/_templates/sdk_TP_tutorial.rst]: These two sentences can be merged into one :
```
You should be familiar with the concepts introduced in the 
:doc:`/app_developers_guide/getting_started` guide and have a working Sawtooth 
environment prior to completing this tutorial.
```

File: docs/source/app_developers_guide/docker.rst [hunk 35]
```diff
@@ -0,0 +1,416 @@
...
+for application development using Docker Compose, and then introduces basic
+Sawtooth concepts necessary for application development.
+
+After following the steps in this document, you will be able to perform the
+following tasks:
+
+* run Sawtooth
+* submit transactions
+* view blocks, transactions and state with the CLI
+* start and stop validators and transaction processors
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: This document walks through the process of setting up Hyperledger Sawtooth
for application development using Docker Compose, introduces some of the basic 
Sawtooth concepts necessary for application development, and walks through
performing the following tasks:
* Submit transactions to the REST API
* View blocks, transactions, and state with the `sawtooth` CLI tool
* Start and stop validators and transaction processors

File: docs/source/app_developers_guide/docker.rst [hunk 36]
```diff
@@ -0,0 +1,416 @@
...
+
+You will then be prepared for the more advanced tutorials that guide you in
+performing app development tasks, such as implementing business logic with
+transaction families and writing clients which use Sawtooth's REST API.
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: Upon completing this tutorial, you will be prepared for the more advanced tutorials that guide you in performing app development tasks, such as implementing business logic with transaction families and writing clients which use Sawtooth's REST API.

File: docs/source/app_developers_guide/docker.rst [hunk 37]
```diff
@@ -0,0 +1,416 @@
...
+  sawtooth-validator-default | Generating /var/lib/sawtooth/genesis.batch
+
+Open a new terminal so we can connect to the client container:
+
+.. code-block:: console
+
+  % docker exec -it sawtooth-client-default bash
+
+The client container is used to run sawtooth CLI commands, which is the usual
+way to interact with  validators or validator networks.
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: extra space between "with" and "validators"

File: docs/source/app_developers_guide/docker.rst [hunk 38]
```diff
@@ -0,0 +1,416 @@
...
+
+Installation
+============
+
+Download the Docker Compose File
+--------------------------------
+
+A Docker Compose file is provided which defines the process for constructing
+a Sawtooth environment, including the container images to download and the
+network settings.
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: A Docker Compose file is provided which defines the process for constructing
a simple Sawtooth environment. This environment includes the following containers:

* A single validator using dev-mode consensus
* A REST-API connected to the validator
* Config, IntegerKey, and XO transaction processors
* A client container for running the CLI tools

The compose file also specifies the container images to download from Docker Hub and the
network settings needed for all the containers to communicate correctly.

File: docs/source/app_developers_guide/docker.rst [hunk 39]
```diff
@@ -0,0 +1,416 @@
...
+
+  % docker-compose -f sawtooth-default.yaml down
+
+
+
+.. note::
+
+  Genesis block, validator and transaction processor startup are handled for you
+  if you're using the Docker workflow. If you want a better understanding of
+  this process, see :doc:`ubuntu`.
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: This would make more sense right after the docker-compose up command is run and could be as simple as "To learn more about the startup process, see: ..."

File: docs/source/app_developers_guide/docker.rst [hunk 40]
```diff
@@ -0,0 +1,416 @@
...
+
+
+Confirming Connectivity 
+=======================
+
+To confirm that a validator is up and running, and reachable from the client
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: Since you just showed how to shut down the environment, it would probably be good to have a reminder to restart the environment here. Or you could move the  "Resetting The Environment" section to the end and put a link where it is now instead.

File: docs/source/app_developers_guide/docker.rst [hunk 41]
```diff
@@ -0,0 +1,416 @@
...
+* tp_xo_python
+
+These transaction processors handle transactions for the config, intkey, and XO transaction families, and are written in Python.
+
+
+Creating And Submitting Transactions
+====================================
+
+The **intkey** command is provided to create sample transactions of the intkey
+transaction type for testing purposes.
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: The **intkey** command can be used to create and submit IntegerKey transactions. The transactions created conform to the :doc:`/transaction_family_specifications/integerkey_transaction_family`

File: docs/source/app_developers_guide/docker.rst [hunk 42]
```diff
@@ -0,0 +1,416 @@
...
+
+This section guide you through the following tasks:
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: "This sections will show you how to:"
(match tense with bullets)

File: docs/source/app_developers_guide/docker.rst [hunk 43]
```diff
@@ -0,0 +1,416 @@
...
+
+1. Prepare a batch of intkey transactions that set the keys to random values.
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: "that set some keys"

File: docs/source/app_developers_guide/docker.rst [hunk 44]
```diff
@@ -0,0 +1,416 @@
...
+
+2. Generate *inc* (increment) and *dec* (decrement) transactions to apply to
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: "to modify existing global state"

File: docs/source/app_developers_guide/docker.rst [hunk 45]
```diff
@@ -0,0 +1,416 @@
...
+
+.. code-block:: console
+
+  $ intkey create_batch
+  $ intkey load -f batches.intkey -U http://rest_api:8080
+
+You can observe the processing of the intkey transactions by observing the
+logging output of the intkey transaction processor and the validator, from the
+terminal window in which you ran the *docker-compose* command. A truncated
+example of the output:
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: The terminal window in which you ran the *docker-compose* command will begin logging output as the validator and intkey transaction processor handle the transactions just submitted.

File: docs/source/app_developers_guide/docker.rst [hunk 46]
```diff
@@ -0,0 +1,416 @@
...
+.. code-block:: console
+
+  sawtooth.config.vote.authorized_keys: 035bd41bf6ea872...
+  sawtooth.validator.transaction_families: [{"family": "in...
+
+
+Viewing Blocks and State
+========================
+
+You can view the blocks stored in the block-chain, and the nodes of the Merkle
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: blockchain

File: docs/source/app_developers_guide/docker.rst [hunk 47]
```diff
@@ -0,0 +1,416 @@
...
+        - 1cf126c15b04cb20206d45c4d0e432d036420401dbd90f064683399fae55b99af1a543f7de79cfafa4f220a22fa248f8346fb1ad0343fcf8d7708565ebb8a3deaac09d
+        payload_encoding: application/cbor
+        payload_sha512: 942a09c0254c4a5712ffd152dc6218fc5453451726d935ac1ba67de93147b5e7be605da7ab91245f48029b41f493a1cc8dfc45bb090ac97420580eb1bdded01f
+        signer_pubkey: 0380be3421629849b1d03af520d7fa2cdc24c2d2611771ddf946ef3aaae216be84
+      header_signature: c498c916da09450597053ada1938858a11d94e2ed5c18f92cd7d34b865af646144d180bdc121a48eb753b4abd326baa3ea26ee8a29b07119052320370d24ab84
+      payload: o2ROYW1lZnFrbGR1emVWYWx1ZQFkVmVyYmNpbmM=
+
+
+
+Viewing The State
```
Comment (aludvik) [inline docs/source/app_developers_guide/docker.rst]: Viewing Global State

File: docs/source/_templates/sdk_TP_tutorial.rst (hunk 34 above)
Comment (ojalatodd) [inline docs/source/_templates/sdk_TP_tutorial.rst]: Good suggestions. Changed.

File: docs/source/app_developers_guide/docker.rst (hunk 35 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: I like it. Used your suggestion. Better flow.

File: docs/source/app_developers_guide/docker.rst (hunk 36 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: good

File: docs/source/app_developers_guide/docker.rst (hunk 38 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Used suggestion

File: docs/source/app_developers_guide/docker.rst (hunk 37 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed

File: docs/source/app_developers_guide/docker.rst (hunk 39 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Moved and shortened.

File: docs/source/app_developers_guide/docker.rst (hunk 40 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Added reminder with links.

File: docs/source/app_developers_guide/docker.rst (hunk 42 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed

File: docs/source/app_developers_guide/docker.rst (hunk 43 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Great catch.

File: docs/source/app_developers_guide/docker.rst (hunk 44 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Changed

File: docs/source/app_developers_guide/docker.rst (hunk 45 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Great suggestion.

File: docs/source/app_developers_guide/docker.rst (hunk 46 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed

Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Also fixed in Ubuntu version

File: docs/source/app_developers_guide/docker.rst (hunk 47 above)
Comment (ojalatodd) [inline docs/source/app_developers_guide/docker.rst]: Fixed here and in Ubuntu version
